embedded-kafka/src/main/scala/io/github/embeddedkafka/ops/kafkaOps.scala

```diff
@@ -1,62 +1,179 @@
 package io.github.embeddedkafka.ops
 
-import java.nio.file.Path
-import kafka.server.{KafkaConfig, KafkaServer}
-import io.github.embeddedkafka.{
-  EmbeddedK,
-  EmbeddedKafkaConfig,
-  EmbeddedServer,
-  EmbeddedZ
-}
+import io.github.embeddedkafka.{EmbeddedK, EmbeddedKafkaConfig, EmbeddedServer}
+import kafka.server._
+import org.apache.kafka.common.Uuid
 import org.apache.kafka.common.security.auth.SecurityProtocol
+import org.apache.kafka.common.utils.Time
 import org.apache.kafka.coordinator.group.GroupCoordinatorConfig
-import org.apache.kafka.coordinator.transaction.TransactionLogConfigs
+import org.apache.kafka.metadata.properties.{
+  MetaProperties,
+  MetaPropertiesEnsemble,
+  MetaPropertiesVersion,
+  PropertiesUtils
+}
 import org.apache.kafka.network.SocketServerConfigs
+import org.apache.kafka.raft.QuorumConfig
+import org.apache.kafka.server.ServerSocketFactory
 import org.apache.kafka.server.config.{
+  KRaftConfigs,
   ServerConfigs,
-  ServerLogConfigs,
-  ZkConfigs
+  ServerLogConfigs
 }
 import org.apache.kafka.storage.internals.log.CleanerConfig
+import org.slf4j.LoggerFactory
 
+import java.io.{File, IOException}
+import java.net.ServerSocket
+import java.nio.file.{Path, Paths}
+import java.util.concurrent.CompletableFuture
 import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Try, Using}
 
 /**
   * Trait for Kafka-related actions.
   */
 trait KafkaOps {
-  protected val brokerId: Short                 = 0
+
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  protected val nodeId: Int                     = 0
   protected val autoCreateTopics: Boolean       = true
   protected val logCleanerDedupeBufferSize: Int = 1048577
 
   private[embeddedkafka] def startKafka(
       kafkaPort: Int,
-      zooKeeperPort: Int,
+      controllerPort: Int,
       customBrokerProperties: Map[String, String],
       kafkaLogDir: Path
-  ) = {
-    val zkAddress = s"localhost:$zooKeeperPort"
-    val listener  = s"${SecurityProtocol.PLAINTEXT}://localhost:$kafkaPort"
-
-    val brokerProperties = Map[String, Object](
-      ZkConfigs.ZK_CONNECT_CONFIG                     -> zkAddress,
-      ServerConfigs.BROKER_ID_CONFIG                  -> brokerId.toString,
-      SocketServerConfigs.LISTENERS_CONFIG            -> listener,
-      SocketServerConfigs.ADVERTISED_LISTENERS_CONFIG -> listener,
+  ): (BrokerServer, ControllerServer) = {
+
+    // We need to know the controller port beforehand to set the config for the broker
+    // Without this the controller starts correctly on a random port but it's too late to use this port in the configs for the broker
+    val actualControllerPort = findPortForControllerOrFail(controllerPort)
+
+    val brokerListener = s"${SecurityProtocol.PLAINTEXT}://localhost:$kafkaPort"
+    val controllerListener = s"CONTROLLER://localhost:$actualControllerPort"
```

Comment:
I have quite some doubt and missing knowledge regarding the controller listener: using `CONTROLLER` as a name works but seems like it should rather be a `SecurityProtocol`. But then if we use `PLAINTEXT` for both the broker and the controller, they have both have the same name.

:thinking: 

If anyone knows a bit more about this, feedback will be greatly appreciated.

---

This seems to cause issues when I test zio-kafka projects that has some SASL tests. Still investigating this...

Reply:
I checked in the official Kafka docker image, it starts in KRaft combined mode (as we want to) with the following properties:
```
advertised.listeners = PLAINTEXT://localhost:9092
controller.listener.names = CONTROLLER
controller.quorum.voters = [1@localhost:9093]
listener.security.protocol.map = CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT,SSL:SSL,SASL_PLAINTEXT:SASL_PLAINTEXT,SASL_SSL:SASL_SSL
listeners = PLAINTEXT://:9092,CONTROLLER://:9093
```

So, what we have so far in embedded-kafka seems adequate.